core/templates/dev/head/pages/exploration_editor/ExplorationSaveService.js

```diff
@@ -0,0 +1,511 @@
+// Copyright 2016 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Functionality for the create exploration button and upload
+ * modal.
+ */
+
+oppia.factory('explorationSaveService', [
+  '$http', '$modal', '$timeout', '$rootScope', '$window', '$log',
+  'alertsService', 'explorationData', 'explorationStatesService',
+  'explorationTagsService', 'explorationTitleService',
+  'explorationObjectiveService', 'explorationCategoryService',
+  'explorationLanguageCodeService', 'explorationRightsService',
+  'explorationWarningsService', 'ExplorationDiffService',
+  'explorationInitStateNameService', 'routerService',
+  'focusService', 'changeListService', 'siteAnalyticsService',
+  function(
+      $http, $modal, $timeout, $rootScope, $window, $log,
+      alertsService, explorationData, explorationStatesService,
+      explorationTagsService, explorationTitleService,
+      explorationObjectiveService, explorationCategoryService,
+      explorationLanguageCodeService, explorationRightsService,
+      explorationWarningsService, ExplorationDiffService,
+      explorationInitStateNameService, routerService,
+      focusService, changeListService, siteAnalyticsService) {
+    // Whether or not a save action is currently in progress.
+    var saveInProgress = false;
+    // Whether or not a discard action is currently in progress.
+    var discardInPrograss = false;
+
+    var publishModalOpening = false;
+
+    // This flag is used to change text of save button to "Loading..." to
+    // add indication for user that something is happening.
+    var saveModalIsOpening = false;
```

Comment:
This should actually go in the controller, since it is specific to that particular UI and not a general construct. Maybe you could have an onFinishOpen() callback (or, better, find a way to return a promise that is resolved when the modal opens so that the controller can act on it).